staging/src/k8s.io/api/autoscaling/v2/types.go

```diff
@@ -0,0 +1,537 @@
+type MetricTargetType string
+
+const (
+	// UtilizationMetricType declares a MetricTarget is an AverageUtilization value
+	UtilizationMetricType MetricTargetType = "Utilization"
+	// ValueMetricType declares a MetricTarget is a raw value
+	ValueMetricType MetricTargetType = "Value"
+	// AverageValueMetricType declares a MetricTarget is an
+	AverageValueMetricType MetricTargetType = "AverageValue"
+)
+
+// HorizontalPodAutoscalerStatus describes the current status of a horizontal pod autoscaler.
+type HorizontalPodAutoscalerStatus struct {
+	// observedGeneration is the most recent generation observed by this autoscaler.
+	// +optional
+	ObservedGeneration *int64 `json:"observedGeneration,omitempty" protobuf:"varint,1,opt,name=observedGeneration"`
+
+	// lastScaleTime is the last time the HorizontalPodAutoscaler scaled the number of pods,
+	// used by the autoscaler to control how often the number of pods is changed.
+	// +optional
+	LastScaleTime *metav1.Time `json:"lastScaleTime,omitempty" protobuf:"bytes,2,opt,name=lastScaleTime"`
+
+	// currentReplicas is current number of replicas of pods managed by this autoscaler,
+	// as last seen by the autoscaler.
+	CurrentReplicas int32 `json:"currentReplicas" protobuf:"varint,3,opt,name=currentReplicas"`
+
+	// desiredReplicas is the desired number of replicas of pods managed by this autoscaler,
+	// as last calculated by the autoscaler.
+	DesiredReplicas int32 `json:"desiredReplicas" protobuf:"varint,4,opt,name=desiredReplicas"`
+
+	// currentMetrics is the last read state of the metrics used by this autoscaler.
+	// +optional
+	CurrentMetrics []MetricStatus `json:"currentMetrics" protobuf:"bytes,5,rep,name=currentMetrics"`
+
+	// conditions is the set of conditions required for this autoscaler to scale its target,
+	// and indicates whether or not those conditions are met.
+	Conditions []HorizontalPodAutoscalerCondition `json:"conditions" protobuf:"bytes,6,rep,name=conditions"`
+}
+
+// HorizontalPodAutoscalerConditionType are the valid conditions of
+// a HorizontalPodAutoscaler.
+type HorizontalPodAutoscalerConditionType string
+
+const (
+	// ScalingActive indicates that the HPA controller is able to scale if necessary:
+	// it's correctly configured, can fetch the desired metrics, and isn't disabled.
+	ScalingActive HorizontalPodAutoscalerConditionType = "ScalingActive"
+	// AbleToScale indicates a lack of transient issues which prevent scaling from occurring,
+	// such as being in a backoff window, or being unable to access/update the target scale.
+	AbleToScale HorizontalPodAutoscalerConditionType = "AbleToScale"
+	// ScalingLimited indicates that the calculated scale based on metrics would be above or
+	// below the range for the HPA, and has thus been capped.
+	ScalingLimited HorizontalPodAutoscalerConditionType = "ScalingLimited"
+)
+
+// HorizontalPodAutoscalerCondition describes the state of
+// a HorizontalPodAutoscaler at a certain point.
+type HorizontalPodAutoscalerCondition struct {
```

Comment:
Can we instead of a generic Condition:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1398

@thockin ?

Reply:
Oh, good idea.  We could use https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Condition.  The only thing we don't have in the HPA conditions is ObservedGeneration.  So we would need to serialize / deserialize that to annotations and populate that in the controller.  Unfortunately the controller is based on the v2beta2 objects, so that would be a chicken and egg problem (we would want to operate on v2 stable in the controller to populate the new field, but need to land it first).

Reply:
We could just add ObservedGeneration to the v2beta2 API.  Would have to happen first.

Reply:
Isn't controller operating on the internal objects? Adding to internal representation is pretty safe.

Reply:
No, the controller is operating on public v2beta2 objects.  :(  I don't know why.

Reply:
All external clients operate on external types.

Reply:
@liggitt I think the point is that adding a new field in the stable API (condition observedGeneration) is more difficult in the case of a controller like HPA which uses an external type (v2beta2) internally.  There is no internal representation of HorizontalPodAutoscaler.  To land this new API we (maybe) need to have the controller add some data to the internal type v2beta2 (condition observedGeneration).  But we don't have anywhere to put it it until we migrate the controller to v2 stable.

Reply:
I think the generic conditions are best-suited for new APIs. Retrofitting them onto a specific version of an API which already uses a local condition type is likely not worth the effort.